config/initializers/services.rb

```diff
@@ -0,0 +1,64 @@
+module SelfService
+
+  def self.register_service(name:, client:)
+    @services ||= {}
+
+    @services[name] = client
+  end
+
+  def self.service(name)
+    @services[name] || raise(ServiceNotRegisteredException.new(name))
+  end
+
+  class ServiceNotRegisteredException < RuntimeError; end
+end
+
+def aws_access_key
+  Rails.application.secrets.cognito_aws_access_key_id
+end
+
+def aws_secret_key
+  Rails.application.secrets.cognito_aws_secret_access_key
+end
+
+def stub_client
+  Rails.application.secrets.cognito_client_id = SecureRandom.uuid
+  client = Aws::CognitoIdentityProvider::Client.new(stub_responses: true)
+  client.stub_responses(:initiate_auth, { challenge_name: nil, authentication_result: {access_token: "valid-token" }})
+  client.stub_responses(:get_user, {username: "2dc8b36a-67c6-4d01-a862-67416e507979", user_attributes:
+    [
+      {name: "sub", value: "2dc8b36a-67c6-4d01-a862-67416e507979"},
+      {name: "custom:roles", value: "dev"},
+      {name: "email_verified", value: "true"},
+      {name: "custom:organisation", value: "GDS"},
+      {name: "phone_number_verified", value: "true"},
+      {name: "phone_number", value: "+447000000000"},
+      {name: "given_name", value: "Test"},
+      {name: "family_name", value: "User"},
+      {name: "email", value: "test@test.test"}
+    ],
+  preferred_mfa_setting: "SOFTWARE_TOKEN_MFA",
+  user_mfa_setting_list: ["SOFTWARE_TOKEN_MFA"]})
+  client
+end
+
+def cognito_client
+  return Aws::CognitoIdentityProvider::Client.new if Rails.env == "production"
+
+  if aws_access_key.present? && aws_secret_key.present? && Rails.env != "test"
```

Comment:
do we need the last condition `Rails.env != "test"`?

Reply:
Yes because if environment variables are set or a .env file is present it will always load up the the real cognito instance which will break all the tests.  So we need to make sure we don't allow the real auth to load at this point

Reply:
OK, I see could shuffle it around like below, but leave it for now.

e.g.
```
...
return Aws::CognitoIdentityProvider::Client.new if Rails.env == "production"
return stub_client if Rails.env == "test"
return Aws::CognitoIdentityProvider::Client.new(local_config) if aws_access_key.present?
return stub_client if Rails.env == "development"

...
def local_config
  {       
      region: Rails.application.secrets.aws_region,
      access_key_id: aws_access_key,
      secret_access_key: aws_secret_key
  }
```

Futhermore, this file probably should be in `/app/services/authentication.rb` so that the initializers are light only calling the right thing... But this can be refactored later on when we make the stub auth smarter...